parsers/c/c_lexer.mll

```diff
@@ -92,70 +92,98 @@ let lexicon: (string, token) Hashtbl.t =
+    "i64"           , (Production, CN_BITS (`I,64));
+    "i128"          , (Production, CN_BITS (`I,128));
+    "real"          , (Production, CN_REAL);
+    "pointer"       , (Production, CN_POINTER);
+    "alloc_id"      , (Production, CN_ALLOC_ID);
+    "map"           , (Production, CN_MAP);
+    "let"           , (Production, CN_LET);
+    "take"          , (Production, CN_TAKE);
+    "Owned"         , (Production, CN_OWNED);
+    "Block"         , (Production, CN_BLOCK);
+    "each"          , (Production, CN_EACH);
+    "NULL"          , (Production, CN_NULL);
+    "true"          , (Production, CN_TRUE);
+    "false"         , (Production, CN_FALSE);
+    "requires"      , (Production, CN_REQUIRES);
+    "ensures"       , (Production, CN_ENSURES);
+    "inv"           , (Production, CN_INV);
+    "accesses"      , (Production, CN_ACCESSES);
+    "trusted"       , (Production, CN_TRUSTED);
+    "spec"          , (Production, CN_SPEC);
+    "unchanged"     , (Production, CN_UNCHANGED);
+    "instantiate"   , (Production, CN_INSTANTIATE);
+    "split_case"    , (Production, CN_SPLIT_CASE);
+    "extract"       , (Production, CN_EXTRACT);
+    "array_shift"   , (Production, CN_ARRAY_SHIFT);
+    "member_shift"  , (Production, CN_MEMBER_SHIFT);
+    "unfold"        , (Production, CN_UNFOLD);
+    "apply"         , (Production, CN_APPLY);
+    "match"         , (Production, CN_MATCH);
+    "predicate"     , (Production, CN_PREDICATE);
+    "function"      , (Production, CN_FUNCTION);
+    "lemma"         , (Production, CN_LEMMA);
+    "datatype"      , (Production, CN_DATATYPE);
+    "type_synonym"  , (Production, CN_TYPE_SYNONYM);
+    "_"             , (Production, CN_WILD);
+    "implies"       , (Production, CN_IMPLIES);
+
+    (* CN 'experimental' keywords - functional in some cases but not recommended for
+    general use *)
+    "cn_list"       , (Experimental, CN_LIST); 
+    "cn_tuple"      , (Experimental, CN_TUPLE);
+    "cn_set"        , (Experimental, CN_SET);
+    "cn_have"       , (Experimental, CN_HAVE);
+    "cn_function"   , (Experimental, CN_FUNCTION);
+    "cn_print"      , (Experimental, CN_PRINT);
+
+    (* CN 'unimplemented' keywords - non-functional, but the keyword is reserved *)
+    "pack"          , (Unimplemented, CN_PACK);
+    "unpack"        , (Unimplemented, CN_UNPACK);
+  ] 
+
+(* Attempt to lex a CN keyword. These may be: 
+  * 'production' - well-supported and suitable for general use
+  * 'experimental' - functional in some cases but not recommended for general use 
+  * 'unimplemented' - non-functional, but the keyword is reserved 
+
+May raise `Not_found`, indicating `id` is not a recognized CN keyword. *)
+let cn_lex_keyword id start_pos end_pos = 
+  (* Try to lex CN production keywords *)
+  let cn_lex = Hashtbl.create 0 in
+  let add (key, builder) = Hashtbl.add cn_lex key builder in
+  List.iter add cn_keywords; 
+  match Hashtbl.find cn_lex id with 
```

Comment:
@cp526 @septract this is recreating the Hashtable on each call to look up a keyword; probably want something like the below...
```ocaml
let cn_lex_keyword = 
  (* Try to lex CN production keywords *)
  let cn_lex = Hashtbl.create 0 in
  let add (key, builder) = Hashtbl.add cn_lex key builder in
  List.iter add cn_keywords;
  fun id start_pos end_pos ->
  match Hashtbl.find cn_lex id with 
  ```